UefiPayloadPkg/Library/AmdSpiLib/AmdSpiLib.c

```diff
@@ -55,15 +89,56 @@ AmdSpiEraseBlock (
 {
   UINT32 address = ADDRESS(Lba, 0);
 
-  return spi_flash_erase(&flash, address, BLOCK_SIZE);
+  return spi_flash_erase(flash, address, BLOCK_SIZE);
 }
 
 EFI_STATUS
 AmdSpiInitialize (VOID)
 {
   DEBUG((EFI_D_INFO, "%a\n", __FUNCTION__));
 
-  spi_init();
+  spi_base = (VOID *)spi_get_bar();
+
+  flash = AllocateRuntimeZeroPool (sizeof(struct spi_flash));
+  if (!flash) {
+    DEBUG((EFI_D_ERROR, "%a: Out of resources\n", __FUNCTION__));
+    return EFI_OUT_OF_RESOURCES;
+  }
+
+  flash->ops = AllocateRuntimeZeroPool (sizeof(struct spi_flash_ops));
+  if (!flash->ops) {
+    DEBUG((EFI_D_ERROR, "%a: Out of resources\n", __FUNCTION__));
+    return EFI_OUT_OF_RESOURCES;
+  }
+
+  flash->spi.ctrlr = AllocateRuntimeZeroPool (sizeof(struct spi_ctrlr));
+  if (!flash->spi.ctrlr) {
+    DEBUG((EFI_D_ERROR, "%a: Out of resources\n", __FUNCTION__));
+    return EFI_OUT_OF_RESOURCES;
+  }
+
+  PcdSet32S (PcdFchSpiBar, (UINT32)(UINTN)spi_base);
 
-  return spi_flash_probe(0, 0, &flash);
+  return spi_flash_probe(0, 0, flash);
+}
+
+VOID
+EFIAPI
+AmdSpiVirtualNotifyEvent (
```

Comment:
@pietrushnic the actual problem is here. It doesn't work yet, probably because not all pointers are converted for runtime virtual address change. Also the AmdSpiLib should have a constructor that will create the virtual address change event with this function and a destructor that will close this event. Yet I think I am missing a lot here.